core/modules/replica/Controller.cc

```diff
@@ -0,0 +1,837 @@
+        onFinish,
+        _messenger,
+        requestExpirationIvalSec);
+}
+
+ServiceStatusRequest::pointer Controller::statusOfWorkerService(
+                                    std::string const& workerName,
+                                    ServiceStatusRequest::callback_type onFinish,
+                                    std::string const& jobId,
+                                    unsigned int requestExpirationIvalSec) {
+    LOCK_GUARD;
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "statusOfWorkerService  workerName: " << workerName);
+
+    return ControllerImpl::serviceManagementOperation<ServiceStatusRequest>(
+        shared_from_this(),
+        jobId,
+        workerName,
+        onFinish,
+        _messenger,
+        requestExpirationIvalSec);
+}
+
+ServiceRequestsRequest::pointer Controller::requestsOfWorkerService(
+                                    std::string const& workerName,
+                                    ServiceRequestsRequest::callback_type onFinish,
+                                    std::string const& jobId,
+                                    unsigned int requestExpirationIvalSec) {
+    LOCK_GUARD;
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "requestsOfWorkerService  workerName: " << workerName);
+
+    return ControllerImpl::serviceManagementOperation<ServiceRequestsRequest>(
+        shared_from_this(),
+        jobId,
+        workerName,
+        onFinish,
+        _messenger,
+        requestExpirationIvalSec);
+}
+
+ServiceDrainRequest::pointer Controller::drainWorkerService(
+                                    std::string const& workerName,
+                                    ServiceDrainRequest::callback_type onFinish,
+                                    std::string const& jobId,
+                                    unsigned int requestExpirationIvalSec) {
+    LOCK_GUARD;
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "drainWorkerService  workerName: " << workerName);
+
+    return ControllerImpl::serviceManagementOperation<ServiceDrainRequest>(
+        shared_from_this(),
+        jobId,
+        workerName,
+        onFinish,
+        _messenger,
+        requestExpirationIvalSec);
+}
+
+size_t Controller::numActiveRequests() const {
+    LOCK_GUARD;
+    return _registry.size();
+}
+
+void Controller::finish(std::string const& id) {
+
+    // IMPORTANT:
+    //
+    //   Make sure the notification is complete before removing
+    //   the request from the registry. This has two reasons:
+    //
+    //   - it will avoid a possibility of deadlocking in case if
+    //     the callback function to be notified will be doing
+    //     any API calls of the controller.
+    //
+    //   - it will reduce the controller API dead-time due to a prolonged
+    //     execution time of of the callback function.
+
+    ControllerRequestWrapper::pointer request;
+    {
+        LOCK_GUARD;
+        request = _registry[id];
+        _registry.erase(id);
+    }
+    request->notify();
+}
```

Comment:
I'm confused: the comment under "IMPORTANT" above says to make sure the notification is complete before removing the request from the registry, but it looks like the code does the opposite; it removes the request from the registry and then does the notification when the request is no longer in the registry. Am I missing something?

Reply:
Yes, indeed. Actually, this mistake had already been noticed and fixed before I had a chance to read your comment. The new version says:
...
Make sure the lock is released before sending notifications:
...